src/compiler/checker.ts

```diff
@@ -3677,11 +3678,25 @@ namespace ts {
                 const additionalContainers = mapDefined(container.declarations, fileSymbolIfFileSymbolExportEqualsContainer);
                 const reexportContainers = enclosingDeclaration && getAlternativeContainingModules(symbol, enclosingDeclaration);
                 const objectLiteralContainer = getVariableDeclarationOfObjectLiteral(container, meaning);
-                if (enclosingDeclaration && getAccessibleSymbolChain(container, enclosingDeclaration, SymbolFlags.Namespace, /*externalOnly*/ false)) {
+                if (enclosingDeclaration && container.flags & getQualifiedLeftMeaning(meaning) && getAccessibleSymbolChain(container, enclosingDeclaration, SymbolFlags.Namespace, /*externalOnly*/ false)) {
                     return append(concatenate(concatenate([container], additionalContainers), reexportContainers), objectLiteralContainer); // This order expresses a preference for the real container if it is in scope
                 }
+                // we potentially a symbol which is a member of the instance side of something - look for a variable in scope with the container's type
+                // which may be acting like a namespace
+                const firstVariableMatch = !(container.flags & getQualifiedLeftMeaning(meaning))
+                    && container.flags & SymbolFlags.Type
+                    && getDeclaredTypeOfSymbol(container).flags & TypeFlags.Object
+                    && meaning === SymbolFlags.Value
+                ? forEachSymbolTableInScope(enclosingDeclaration, t => {
+                    return forEachEntry(t, s => {
+                        if (s.flags & getQualifiedLeftMeaning(meaning) && getTypeOfSymbol(s) === getDeclaredTypeOfSymbol(container)) {
+                            return s;
+                        }
+                    });
+                }) : undefined;
                 const res = append(append(additionalContainers, container), objectLiteralContainer);
-                return concatenate(res, reexportContainers);
+                const resWithReexports = concatenate(res, reexportContainers);
+                return firstVariableMatch ? [firstVariableMatch, ...resWithReexports] : resWithReexports;

```

Comment:
I don't think a bunch of `.push` or `.unshift` calls make it that much nicer. Plus, it makes the final ordering less clear, imo.

Reply:
It might not be too bad

```ts
const res = firstVariableMatch ? [firstVariableMatch, ...additionalContainers] : additionalContainers;
append(res, container);
append(res, objectLiteralContainer);
addRange(res, reexportContainers);
return res;
```

I guess the thing you lose is the distinction between which ones are re-exports?

Reply:
Alternatively

```ts
return (emptyArray as Node[]).concat(
    firstVariableMatch,
    additionalContainers,
    container,
    objectLiteralContainer,
    reexportContainers,
);
```

which might be able to pre-allocate.

Reply:
For the later, it doesn't handle the possible-`undefined`-ness that many of those members have, which is the entire reason `concatenate` is in use here - it handles `undefined`. `reexportContainers` is possibly `undefined`, which means that `concat` call would add an `undefined` to the end of the array, which would be bad.